### PR TITLE
App by id Endpoint Implementation

### PR DESCRIPTION
**Pull request by janiejaffe**

### CODE REVIEW

File: handlers/app.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+import (
+
+	"github.com/knative-scout/app-api/models"
+	"go.mongodb.org/mongo-driver/bson"
+	"net/http"
+	"github.com/gorilla/mux"
+)
+
+// HealthHandler is used to determine if the server is running
```
Comment (catvec) [inline handlers/app.go]: Can you make the doc comment match what your handler does?

Something like:

```
// AppHandler returns a single app by ID from the database
```

File: handlers/app.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+type AppHandler struct {
```
Comment (catvec) [inline handlers/app.go]: Can you rename this handler to something less general? B/c we will have a few more app endpoints.

Maybe name it `GetAppByIDHandler`?

File: handlers/app.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+
+func getDataFromDB(id string, h AppHandler ) models.App{
+
+	ret := []models.App{}
+	result, err := h.MDbApps.Find(h.Ctx, bson.D{{"app_id", id}})
+
+	if err != nil {
+		 h.Logger.Fatalf("failed to retrieve data from db %s", err)
```
Comment (catvec) [inline handlers/app.go]: Can you `panic()` the errors instead?

```
panic(fmt.Errorf("failed to retrieve data from db: %s", err.Error()))
```

File: handlers/app.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	for result.Next(h.Ctx) {
+		a := models.App{}
+		if err = result.Decode(&a); err != nil {
+			h.Logger.Fatalf("readTasks: couldn't make to-do item ready for display: %v", err)
```
Comment (catvec) [inline handlers/app.go]: ditto.

Also be sure to add errors into the string by using `%s` and calling `err.Error()`

**Review by catvec — CHANGES_REQUESTED** [review]
Good first draft, requested a few minor changes, if you could also rename the `app.go` file to `getapp.go` or something.

File: handlers/app.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+	"github.com/knative-scout/app-api/models"
+	"go.mongodb.org/mongo-driver/bson"
+	"net/http"
+	"github.com/gorilla/mux"
+	"fmt"
+)
+
+// AppByIDHandler returns a single app by ID from the database
+type AppByIDHandler struct {
```
Comment (catvec) [inline handlers/app.go]: Can you rename the file? From `handlers/app.go` to something like `handlers/appbyid.go`

**Comment by janiejaffe** [comment]
I didn't rename it, figure we'll probably roll this into apps.go once that is up

File: handlers/app.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+func (h AppByIDHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	//gets all the optional parametres passed in the URL
+	vars := mux.Vars(r)
+
+	id := vars["id"]
+
+	resp := getDataFromDB(id, h)
+
+	h.RespondJSON(w, http.StatusOK, resp)
```
Comment (catvec) [inline handlers/app.go]: Can you please make the endpoint return data in the format which the spec documented? https://github.com/knative-scout/app-api/blob/master/DESIGN.md#get-app-by-id

It should return an object with an `app` key

File: handlers/app.go [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+
+func getDataFromDB(id string, h AppByIDHandler ) models.App{
```
Comment (catvec) [inline handlers/app.go]: If you go to `localhost:5000/apps/this-app-id-does-not-exist` your program will crash.

Can you make it so the endpoint responds with `http.StatusNotFound` and a response of `{"error": "not found"}`?

**Review by catvec — CHANGES_REQUESTED** [review]
Working well, found a common edge case where it crashes though.

**Review by catvec — CHANGES_REQUESTED** [review]
Can't run, crashes with:

```
% go run .
# github.com/knative-scout/app-api/handlers
handlers/appbyid.go:25:22: not enough arguments in call to getDataFromDB
        have (string, AppByIDHandler)
        want (string, string, string, AppSearchHandler)
handlers/search.go:28:25: too many arguments in call to getDataFromDB
        have (string, string, string, AppSearchHandler)
        want (string, AppByIDHandler)
handlers/search.go:31:3: cannot use result (type *models.App) as type []models.App in map value
handlers/search.go:39:6: getDataFromDB redeclared in this block
        previous declaration at handlers/appbyid.go:38:51
```

**Review by catvec — APPROVED** [review]

**Pull request merged**